# Peer-graded Assignment Part B

## Instructions

B. Normalize the data (5 marks)

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [1]:
# import all required libraries

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from sklearn.metrics import mean_squared_error

#import data

concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')

#Normalize and split the data into training/test sets

y = concrete_data.Strength
X = concrete_data.drop('Strength', axis=1) 
X_norm = (X - X.mean()) / X.std()
n_cols = X_norm.shape[1]

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.3)

Using TensorFlow backend.
/home/nate/.local/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/nate/.local/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/nate/.local/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/nate/.local/lib/python3.7/site-packages/tensorflow/python/framewo

In [2]:
# function to build our model, with 1 hidden layer that has 10 nodes, one output, using the adam optimizer and mean_squared_error for the loss function.

def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# function to build a trained model

def trained_model():
    model = regression_model()
    model.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=0)
    return model

#Build a model and print the mse
model = trained_model()
y_predicted = model.predict(X_test)
mse = mean_squared_error(y_test, y_predicted)
print(mse)


520.9014723894531


In [3]:
# function to train a model and return its MSE
def mse():
    model = trained_model()
    y_predicted = model.predict(X_test)
    mse = mean_squared_error(y_test, y_predicted)
    return mse

# build list of MSEs
# note: this takes a minute.

z = []
for i in range(50):
    z.append(mse())
    #print(z[i])

# mean of z

mn = np.mean(z)
print("Mean: ", mn)

# standard deviation of z

sd = np.std(z)
print("Standard Deviation: ", sd)

Mean:  696.7189404593732
Standard Deviation:  144.11727330401794


Results: 
Mean:  696.7189404593732
Standard Deviation:  144.11727330401794

From part A: Mean:  516.1412238566875
Standard Deviation:  710.1154005722235